src/acl/Asn.cc

```diff
@@ -296,13 +283,21 @@ asHandleReply(void *data, StoreIOBuffer result)
         return;
     }
 
+    if (result.length == 0) {
+        if (result.flags.eof)
+            delete asState; // done
+        else
+            storeClientCopy(asState->sc, e, asState->storeReadBuffer.legacyReadRequest(result.offset), asHandleReply, asState);
+        return;
+    }
+
     /*
      * Next, attempt to parse our request
      * Remembering that the actual buffer size is retsize + reqofs!
      */
     s = buf;
 
-    while ((size_t)(s - buf) < result.length + asState->reqofs && *s != '\0') {
+    while ((size_t)(s - buf) < result.length && *s != '\0') {
```

Comment:
AFAICT, this NUL check and probably another NUL check in the `if` statement below implicitly relied on zeroed ASState::reqbuf (and on result.length being smaller than the buffer size). They should not have, but I feel that _we_ are now responsible for fixing that part of the parsing code because we have to deal with cases where the returned StoreIOBuffer is full. I hope the fix is as simple as replacing the NUL checks with correct result.length checks and not expecting NUL-termination in the Store-returned buffer (including inside asnAddNet()).

AFAICT, it is easy to adjust asnAddNet() to not expect NUL-termination because that function truncates the passed string itself. You will need to pass it the correct non-terminated string length to be used in memchr() call. Please note that _this_ loop cannot continue to terminate `s` because we cannot (continue to incorrectly) assume there is enough buffer space to do so.

The WHOIS [protocol](https://datatracker.ietf.org/doc/html/rfc954) does not guarantee CR, LF, or NUL presence in _responses_. In this PR, we should probably continue to assume that a response may end with a "word". However, be careful with _not_ calling asnAddNet() twice when a word is _split_ between two Store copy() results -- we can only call asnAddNet() when we find a whitespace or EOF.

Reply:
BTW, when you fix a bug, please mention that fix in the commit message, so that we can eventually mention it in the PR description as well.

Reply:
>  Please note that this loop cannot continue to terminate s because we cannot (continue to incorrectly) assume there is enough buffer space to do so.

We can still terminate in the loop if we are within the buffer. If the found word goes up to the end of the buffer - we place it into the leftovers (MemBuf now), without terminating. Also we will need an extra asnAddNet() call at the end of asHandleReply() to parse these leftovers if EOF is observed. In this case, I think we can do without adjusting asnAddNet() - because it will always get null-terminated string.

Reply:
Adjusted at 4e4788b according to my proposal.

> please mention that fix in the commit message,

Mentioned.